dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+
+    Parameters
+    ----------
+    static: array, shape (R, C)
+        static image
+    moving: array, shape (R, C)
+        moving image
+    smask: array, shape (R, C)
+        mask of static object being registered (a binary array with 1's inside
+        the object of interest and 0's along the background)
+    mmask: array, shape (R, C)
+        mask of moving object being registered (a binary array with 1's inside
+        the object of interest and 0's along the background)
+    smin: float
+        the minimum observed intensity associated to the static image, which
+        was used to define the joint PDF
+    sdelta: float
+        bin size associated to the intensities of the static image
+    mmin: float
+        the minimum observed intensity associated to the moving image, which
+        was used to define the joint PDF
+    mdelta: float
+        bin size associated to the intensities of the moving image
+    nbins: int
+        number of histogram bins
+    padding: int
+        number of bins used as padding (the total bins used for padding at both
+        sides of the histogram is actually 2*padding)
+    joint: array, shape (nbins, nbins)
+        the array to write the joint PDF
+    smarginal: array, shape (nbins,)
+        the array to write the marginal PDF associated to the static image
+    mmarginal: array, shape (nbins,)
+        the array to write the marginal PDF associated to the moving image
+    '''
+    cdef:
+        int nrows = static.shape[0]
+        int ncols = static.shape[1]
+        int offset, valid_points
+        cnp.npy_intp i, j, r, c
+        double rn, cn
+        double val, spline_arg, sum
+
+    joint[...] = 0
+    sum = 0
+    valid_points = 0
+    with nogil:
+        smarginal[:] = 0
+        for i in range(nrows):
+            for j in range(ncols):
+                if smask is not None and smask[i, j] == 0:
+                    continue
+                if mmask is not None and mmask[i, j] == 0:
+                    continue
+                valid_points += 1
+                rn = _bin_normalize(static[i, j], smin, sdelta)
+                r = _bin_index(rn, nbins, padding)
+                cn = _bin_normalize(moving[i, j], mmin, mdelta)
+                c = _bin_index(cn, nbins, padding)
+                spline_arg = (c-2) - cn
+
+                smarginal[r] += 1
+                for offset in range(-2, 3):
+                    val = _cubic_spline(spline_arg)
+                    joint[r, c + offset] += val
+                    sum += val
+                    spline_arg += 1.0
+        if sum > 0:
+            for i in range(nbins):
+                for j in range(nbins):
+                    joint[i, j] /= valid_points
+
+            for i in range(nbins):
+                smarginal[i] /= valid_points
+
+            for j in range(nbins):
+                mmarginal[j] = 0
+                for i in range(nbins):
+                    mmarginal[j] += joint[i, j]
+
```

Comment:
Removed extra spaces. Sorry, I understand that it must be annoying to find so many PEP8 errors, so I did a full review using Spyder, as suggested.